### PR TITLE
feat: Add option to pass in the runs-on attribute to enable self-hosted runner

### PR DESCRIPTION
**Pull request by FahadMustafa**
Adds the option to pass in a `runsOn` value to override the default `ubuntu-latest` value.
This is mainly done to enable `self-hosted` runners

### CODE REVIEW

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:110]: lets make this type an enum-like class `Runner` that looks something like this. I'm modeling this off of the aws-cdk lambda [`Runtime`](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/runtime.ts) that we have. I haven't fully thought this through but I'd like it to be modeled something like this:

```ts
export class RuntimeType {
  public static readonly GITHUB_HOSTED = 'github-hosted';
  public static readonly SELF_HOSTED = 'self-hosted';
}

export class Runtime {
  public static readonly UBUNTU_LATEST =  new Runner(['ubuntu-latest'], RuntimeType.GITHUB_HOSTED);

  public static readonly WINDOWS_LATEST = new Runner(['windows-latest'], RuntimeType.GITHUB_HOSTED);

  public static readonly MACOS_LATEST = new Runner(['macos-latest'], RuntimeType.GITHUB_HOSTED);

  public static selfHosted(labels: string[]): Runner {
    return new Runtime(['self-hosted', ...labels]);
  } 

  public constructor(public readonly runsOn: string) {}
}
```

One thing we'll have to figure out is how to get `runsOn: ubuntu-latest` most of the time, but if you want a self-hosted runtime you can have a [list](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/runtime.ts) of labels. Again, I haven't fully thought this through, so feel free to make some changes to the model I'm suggesting.

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Hi @FahadMustafa!

First of all, thanks so much for opening this pull request! This is definitely something we want. In addition to building out the runtime type, please also add a section in the readme detailing what you added.

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: Thanks for the input @kaizen3031593. I've updated the PR.

File: src/workflows-model.ts
Comment (kaizen3031593) [inline src/workflows-model.ts:230]: Let's make the constructor private and get rid of `RunnerType`

File: src/workflows-model.ts
Comment (kaizen3031593) [reply src/workflows-model.ts:235]: If this does not work (maybe there's some JSII error I'm forgetting about for `runsOn`) then lets just always return this.labels.

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Hi @FahadMustafa! Thanks again for opening this PR! For the next review, please do the following:

- decouple `publishAssetsRegion` from this PR and focus on the `runsOn` property
- remove all the formatting changes. they are cluttering the PR diff and make it hard to see what really has changed, and I find them unnecessary. if you really want to change the formatting of the code, please open a new PR for that alone and we can discuss there.
- take a look at my comments on how we should model the `Runner` class.

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: Done. Will raise another PR after this is merged.

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: Removed altogether

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts]: this is not a `string []`

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts]: Done

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts:222]: Updated. Had to return `this.labels`.

**Comment by kaizen3031593** [comment]
@FahadMustafa please also revert the formatting changes that are cluttering the PR. I do not find them necessary and if you disagree, I am happy to hold that discussion on a separate PR. Let's keep this PR to exactly what is necessary for the runs-on attribute

**Comment by FahadMustafa** [comment]
> @FahadMustafa please also revert the formatting changes that are cluttering the PR. I do not find them necessary and if you disagree, I am happy to hold that discussion on a separate PR. Let's keep this PR to exactly what is necessary for the runs-on attribute

@kaizen3031593 apologies for the formatting issues. I don't disagree with any of the current formatting. vscode was auto-formatting on save.

File: src/workflows-model.ts
Comment (kaizen3031593) [inline src/workflows-model.ts:206]: All these public APIs need docs. Specifically for `selfHosted`, please mention that you can supply `self-hosted` as a label, but you do not have to.

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:408]:
```suggestion
        runsOn: this.Runner.runsOn,
```

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:146]:
```suggestion

```

File: README.md
Comment (kaizen3031593) [inline README.md:290]:
```suggestion
The default is `Runner.UBUNTU_LATEST`. You can override this as shown below:
```

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:167]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:174]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:192]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:208]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:257]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:261]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:294]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:310]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:324]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:336]: revert formatting

File: src/pipeline.ts
Comment (kaizen3031593) [inline src/pipeline.ts:343]: revert formatting

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
This is almost there! There's still a lot of formatting changes I'd like you to revert in `src/pipeline.ts`. I tried commenting on a few but there are really a lot.

The last thing I want to think about is whether or not it is worth it to somehow turn
```yaml
runsOn:
  - ubuntu-latest
```
into 
```yaml
runsOn: ubuntu-latest
```
Ideally, we do not disrupt templates like this when they upgrade `cdk-pipelines-github` since its alarming when your template changes and your code does not change. In this case it's probably fine. But something to think about.

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts]: Updated

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts:219]: Added comment to code and mentioned in README

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: updated

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: removed

File: src/pipeline.ts
Comment (FahadMustafa) [reply src/pipeline.ts]: done

File: README.md
Comment (FahadMustafa) [reply README.md]: updated

**Comment by FahadMustafa** [comment]
@kaizen3031593 Somehow I completely missed reverting the formatting on `pipeline.ts`. This is now done. Apologies!

**Comment by FahadMustafa** [comment]
> This is almost there! There's still a lot of formatting changes I'd like you to revert in `src/pipeline.ts`. I tried commenting on a few but there are really a lot.
> 
> The last thing I want to think about is whether or not it is worth it to somehow turn
> 
> ```yaml
> runsOn:
>   - ubuntu-latest
> ```
> 
> into
> 
> ```yaml
> runsOn: ubuntu-latest
> ```
> 
> Ideally, we do not disrupt templates like this when they upgrade `cdk-pipelines-github` since its alarming when your template changes and your code does not change. In this case it's probably fine. But something to think about.

I've made the change to only generate `runsOn` as an array for self-hosted runners.

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
This is near the finish line -- thanks for all your help so far!

Give me until early next week to try this out locally, to make sure the snapshots actually do deploy and succeed (it would help my confidence if you did so as well and report back)!

File: src/workflows-model.ts
Comment (FahadMustafa) [reply src/workflows-model.ts:13]: yes, agree. The alternative is to always render out an array. I feel this is an acceptable trade-off for better end-user experience.

**Comment by FahadMustafa** [comment]
> This is near the finish line -- thanks for all your help so far!
> 
> Give me until early next week to try this out locally, to make sure the snapshots actually do deploy and succeed (it would help my confidence if you did so as well and report back)!

Well, we've actually used this to deploy a static website to S3 on self-hosted runners. We've referenced the source code directly to unblock ourselves. It's working well so far!

**Review by kaizen3031593 — APPROVED** [review]
Thanks for the contribution! This looks great.

**Pull request merged**
